stats/postgres_data_access.go

```diff
@@ -0,0 +1,186 @@
+}
+
+// PgStatsDataAccess implements the StatsDataAccess interface using a Postgres database.
+type PgStatsDataAccess struct {
+	DB *sql.DB
+}
+
+// NewPgStatsDataAccess returns a new instance of PgStatsDataAccess.
+func NewPgStatsDataAccess(db *sql.DB) *PgStatsDataAccess {
+	return &PgStatsDataAccess{DB: db}
+}
+
+// AddPatientStat increments the relevant county and subdivision statistics
+// based on the given patient's address.
+func (da *PgStatsDataAccess) AddPatientStat(patient *models.Patient) (err error) {
+
+	countyfp, cousubfp, err := da.identifyCountyAndSubdivisionForPatient(patient)
+
+	if err != nil {
+		return
+	}
+
+	if patientGenderIsValid(patient) {
+		err = da.updateStats(countyfp, cousubfp, patient.Gender, "increment")
+
+	} else {
+		err = InvalidGenderError
+	}
+	return
+}
+
+// RemovePatientStat decrements the relevant county and subdivision statistics
+// based on the given patient's address.
+func (da *PgStatsDataAccess) RemovePatientStat(patient *models.Patient) (err error) {
+
+	countyfp, cousubfp, err := da.identifyCountyAndSubdivisionForPatient(patient)
+
+	if err != nil {
+		return
+	}
+
+	if patientGenderIsValid(patient) {
+		err = da.updateStats(countyfp, cousubfp, patient.Gender, "decrement")
+
+	} else {
+		err = InvalidGenderError
+	}
+	return
+}
+
+// AddConditionStat increments the relevant county and subdivision condition facts
+// based on the given patient's address.
+func (da *PgStatsDataAccess) AddConditionStat(patient *models.Patient, condition *models.Condition) (err error) {
+
+	countyfp, cousubfp, err := da.identifyCountyAndSubdivisionForPatient(patient)
+	diseasefp, err := da.identifyDiseaseForCondition(condition)
+
+	if err != nil {
+		return
+	}
+
+	if patientGenderIsValid(patient) {
+		err = da.updateFacts(countyfp, cousubfp, diseasefp, patient.Gender, "increment")
+
+	} else {
+		err = InvalidGenderError
+	}
+	return
+}
+
+// RemoveConditionStat decrements the relevant county and subdivision condition facts
+// based on the given patient's address.
+func (da *PgStatsDataAccess) RemoveConditionStat(patient *models.Patient, condition *models.Condition) (err error) {
+
+	countyfp, cousubfp, err := da.identifyCountyAndSubdivisionForPatient(patient)
+	diseasefp, err := da.identifyDiseaseForCondition(condition)
+
+	if err != nil {
+		return
+	}
+
+	if patientGenderIsValid(patient) {
+		err = da.updateFacts(countyfp, cousubfp, diseasefp, patient.Gender, "decrement")
+
+	} else {
+		err = InvalidGenderError
+	}
+	return
+}
+
+// identifyCountyAndSubdivisionForPatient returns the countyfp and cousubfp that
+// match the subdivision in the given patient's address.
+func (da *PgStatsDataAccess) identifyCountyAndSubdivisionForPatient(patient *models.Patient) (countyfp, cousubfp string, err error) {
+	err = da.DB.QueryRow("SELECT countyfp, cousubfp FROM tiger.cousub WHERE name = $1", patient.Address[0].City).Scan(&countyfp, &cousubfp)
```

Comment:
I see, I added some extra error handling to catch this. If the patient's city is valid and somewhere in MA then the query will only return 1 row, guaranteed.